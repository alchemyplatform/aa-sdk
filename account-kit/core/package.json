{
  "name": "@account-kit/core",
  "version": "4.3.1",
  "description": "Core library for account kit that provides state management and framework indepednent abstractions across infra, Alchemy Signer, and Smart Contracts",
  "author": "Alchemy",
  "license": "MIT",
  "private": false,
  "type": "module",
  "main": "./dist/esm/index.js",
  "module": "./dist/esm/index.js",
  "types": "./dist/types/index.d.ts",
  "typings": "./dist/types/index.d.ts",
  "sideEffects": false,
  "files": [
    "dist",
    "src/**/*.ts",
    "!dist/**/*.tsbuildinfo",
    "!vitest.config.ts",
    "!.env",
    "!src/**/*.test.ts",
    "!src/__tests__/**/*"
  ],
  "exports": {
    ".": {
      "types": "./dist/types/index.d.ts",
      "import": "./dist/esm/index.js",
      "default": "./dist/esm/index.js"
    },
    "./package.json": "./package.json"
  },
  "scripts": {
    "prebuild": "npx tsx ./inject-version.ts",
    "build": "yarn clean && yarn build:esm && yarn build:types",
    "build:esm": "tsc --project tsconfig.build.json --outDir ./dist/esm",
    "build:types": "tsc --project tsconfig.build.json --declarationDir ./dist/types --emitDeclarationOnly --declaration --declarationMap",
    "clean": "rm -rf ./dist",
    "docs:gen": "node ../../doc-gen/dist/esm/cli.js generate --in ./src/index.ts --out ../../site/pages/reference/account-kit/core",
    "test": "vitest --passWithNoTests",
    "test:run": "vitest run --passWithNoTests"
  },
  "devDependencies": {
    "@types/js-cookie": "^3.0.6",
    "typescript": "^5.0.4",
    "typescript-template": "*",
    "vitest": "^2.0.4"
  },
  "dependencies": {
    "@account-kit/infra": "^4.3.1",
    "@account-kit/logging": "^4.3.1",
    "@account-kit/signer": "^4.3.1",
    "@account-kit/smart-contracts": "^4.3.1",
    "js-cookie": "^3.0.5",
    "zod": "^3.22.4",
    "zustand": "^5.0.0-rc.2"
  },
  "peerDependencies": {
    "viem": "^2.20.0",
    "wagmi": "^2.12.7"
  },
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/alchemyplatform/aa-sdk.git"
  },
  "bugs": {
    "url": "https://github.com/alchemyplatform/aa-sdk/issues"
  },
  "homepage": "https://github.com/alchemyplatform/aa-sdk#readme",
  "gitHead": "ee46e8bb857de3b631044fa70714ea706d9e317d",
  "optionalDependencies": {
    "alchemy-sdk": "^3.0.0"
  }
}
