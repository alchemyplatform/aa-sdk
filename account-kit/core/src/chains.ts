import { SOLANA_CHAIN_SYMBOL, SOLANA_DEVNET_CHAIN_SYMBOL } from "./symbols.js";
import type { Web3Connection } from "./types.js";

export const SOLANA_CHAIN: Web3Connection["chain"] = {
  id: SOLANA_CHAIN_SYMBOL,
  name: "Solana",
  nativeCurrency: {
    name: "Solana",
    symbol: "SOL",
    decimals: 9,
  },
};

export const SOLANA_DEV_CHAIN: Web3Connection["chain"] = {
  id: SOLANA_DEVNET_CHAIN_SYMBOL,
  name: "Solana Devnet",
  nativeCurrency: {
    name: "Solana",
    symbol: "SOL",
    decimals: 9,
  },
};
