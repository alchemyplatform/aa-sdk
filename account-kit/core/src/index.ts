export type * from "./actions/createAccount.js";
export { createAccount } from "./actions/createAccount.js";
export { disconnect } from "./actions/disconnect.js";
export type * from "./actions/getAccount.js";
export { getAccount } from "./actions/getAccount.js";
export type * from "./actions/getBundlerClient.js";
export { getBundlerClient } from "./actions/getBundlerClient.js";
export { getChain } from "./actions/getChain.js";
export type * from "./actions/getConnection.js";
export { getConnection } from "./actions/getConnection.js";
export type * from "./actions/getSigner.js";
export { getSigner } from "./actions/getSigner.js";
export type * from "./actions/getSignerStatus.js";
export { getSignerStatus } from "./actions/getSignerStatus.js";
export type * from "./actions/getSmartAccountClient.js";
export { getSmartAccountClient } from "./actions/getSmartAccountClient.js";
export type * from "./actions/getUser.js";
export { getUser } from "./actions/getUser.js";
export type * from "./actions/reconnect.js";
export { reconnect } from "./actions/reconnect.js";
export { setChain } from "./actions/setChain.js";
export type * from "./actions/watchAccount.js";
export { watchAccount } from "./actions/watchAccount.js";
export type * from "./actions/watchBundlerClient.js";
export { watchBundlerClient } from "./actions/watchBundlerClient.js";
export { watchChain } from "./actions/watchChain.js";
export type * from "./actions/watchConnection.js";
export { watchConnection } from "./actions/watchConnection.js";
export type * from "./actions/watchSigner.js";
export { watchSigner } from "./actions/watchSigner.js";
export type * from "./actions/watchSignerStatus.js";
export { watchSignerStatus } from "./actions/watchSignerStatus.js";
export type * from "./actions/watchSmartAccountClient.js";
export { watchSmartAccountClient } from "./actions/watchSmartAccountClient.js";
export type * from "./actions/watchUser.js";
export { watchUser } from "./actions/watchUser.js";
export type * from "./createConfig.js";
export { DEFAULT_IFRAME_CONTAINER_ID, createConfig } from "./createConfig.js";
export { ClientOnlyPropertyError } from "./errors.js";
export { hydrate } from "./hydrate.js";
export { defaultAccountState } from "./store/store.js";
export type { SignerStatus } from "./store/types.js";
export type * from "./types.js";
export {
  cookieStorage,
  cookieToInitialState,
  parseCookie,
} from "./utils/cookies.js";

/** Re-export core packages to make it easier to use this package without the lower level packages */
// Accounts
export type * from "@account-kit/smart-contracts";
export {
  MultiOwnerPlugin,
  MultisigPlugin,
  MultisigPluginAbi,
  MultisigPluginExecutionFunctionAbi,
  SessionKeyAccessListType,
  SessionKeyPermissionsBuilder,
  SessionKeyPlugin,
  SessionKeyPluginAbi,
  SessionKeyPluginExecutionFunctionAbi,
  SessionKeySigner,
  accountLoupeActions,
  createLightAccount,
  createMultiOwnerModularAccount as createModularAccount,
  createMultiOwnerLightAccount,
  createMultisigModularAccount,
  getMAInitializationData,
  getMSCAUpgradeToData,
  lightAccountClientActions,
  multiOwnerLightAccountClientActions,
  multiOwnerPluginActions,
  multisigPluginActions,
  multisigSignatureMiddleware,
  pluginManagerActions,
  sessionKeyPluginActions,
} from "@account-kit/smart-contracts";

// Infra
export type * from "@account-kit/infra";
export {
  createAlchemySmartAccountClient,
  type AlchemySmartAccountClient,
  type AlchemySmartAccountClientConfig,
} from "@account-kit/infra";
