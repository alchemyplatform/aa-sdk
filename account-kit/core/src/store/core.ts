import { createAlchemyPublicRpcClient } from "@account-kit/infra";
import type { Chain } from "viem";
import {
  createJSONStorage,
  persist,
  subscribeWithSelector,
} from "zustand/middleware";
import { createStore } from "zustand/vanilla";
import type { Connection } from "../types.js";
import { bigintMapReplacer } from "../utils/replacer.js";
import { bigintMapReviver } from "../utils/reviver.js";
import {
  DEFAULT_STORAGE_KEY,
  type CoreState,
  type CoreStore,
} from "./types.js";

export type CreateCoreStoreParams = {
  connections: Connection[];
  chain: Chain;
  storage?: Storage;
  ssr?: boolean;
};

/**
 * Create the core store for alchemy accounts. This store contains the bundler client
 * as well as the chain configs (including the initial chain to use)
 *
 * @param params connections configs
 * @param params.connections a collection of chains and their connection configs
 * @param params.chain the initial chain to use
 * @param params.storage the storage to use for persisting the state
 * @param params.ssr whether the store is being created on the server
 * @returns the core store
 */
export const createCoreStore = (params: CreateCoreStoreParams): CoreStore => {
  const {
    connections,
    chain,
    storage = typeof window !== "undefined" ? localStorage : undefined,
    ssr,
  } = params;

  // State defined in here should work either on the server or on the client
  // bundler client for example can be used in either setting to make RPC calls
  const coreStore = createStore(
    subscribeWithSelector(
      storage
        ? persist(() => createInitialCoreState(connections, chain), {
            name: `${DEFAULT_STORAGE_KEY}:core`,
            storage: createJSONStorage<CoreState>(() => storage, {
              replacer: (key, value) => {
                if (key === "bundlerClient") {
                  const client = value as CoreState["bundlerClient"];
                  return {
                    connection: connections.find(
                      (x) => x.chain.id === client.chain.id
                    ),
                  };
                }
                return bigintMapReplacer(key, value);
              },
              reviver: (key, value) => {
                if (key === "bundlerClient") {
                  const connection = value as Connection;
                  return createAlchemyPublicRpcClient({
                    chain: connection.chain,
                    connectionConfig: connection,
                  });
                }

                return bigintMapReviver(key, value);
              },
            }),
            version: 1,
            skipHydration: ssr,
          })
        : () => createInitialCoreState(connections, chain)
    )
  );

  return coreStore;
};

const createInitialCoreState = (
  connections: Connection[],
  chain: Chain
): CoreState => {
  const connectionMap = connections.reduce((acc, connection) => {
    acc.set(connection.chain.id, connection);
    return acc;
  }, new Map<number, Connection>());

  if (!connectionMap.has(chain.id)) {
    throw new Error("Chain not found in connections");
  }

  const bundlerClient = createAlchemyPublicRpcClient({
    chain,
    connectionConfig: connectionMap.get(chain.id)!,
  });

  return {
    bundlerClient,
    chain,
    connections: connectionMap,
  };
};
