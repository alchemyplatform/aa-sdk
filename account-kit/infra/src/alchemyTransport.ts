import {
  ChainNotFoundError,
  ConnectionConfigSchema,
  split,
  type ConnectionConfig,
  type NoUndefined,
} from "@aa-sdk/core";
import {
  createTransport,
  http,
  type EIP1193RequestFn,
  type HttpTransportConfig,
  type PublicRpcSchema,
  type Transport,
  type TransportConfig,
} from "viem";
import type { AlchemyRpcSchema } from "./client/types.js";
import { AlchemyChainSchema } from "./schema.js";
import { VERSION } from "./version.js";

type Never<T> = T extends object
  ? {
      [K in keyof T]?: never;
    }
  : never;

type SplitTransportConfig = {
  alchemyConnection: ConnectionConfig;
  nodeRpcUrl: string;
};

const alchemyMethods = [
  "eth_sendUserOperation",
  "eth_estimateUserOperationGas",
  "eth_getUserOperationReceipt",
  "eth_getUserOperationByHash",
  "eth_supportedEntryPoints",
  "rundler_maxPriorityFeePerGas",
  "pm_getPaymasterData",
  "pm_getPaymasterStubData",
];

export type AlchemyTransportConfig = (
  | (ConnectionConfig & Never<SplitTransportConfig>)
  | (SplitTransportConfig & Never<ConnectionConfig>)
) & {
  /** The max number of times to retry. */
  retryCount?: TransportConfig["retryCount"] | undefined;
  /** The base delay (in ms) between retries. */
  retryDelay?: TransportConfig["retryDelay"] | undefined;
  fetchOptions?: NoUndefined<HttpTransportConfig["fetchOptions"]>;
};

type AlchemyTransportBase = Transport<
  "alchemy",
  {
    alchemyRpcUrl: string;
    fetchOptions?: AlchemyTransportConfig["fetchOptions"];
  },
  EIP1193RequestFn<[...PublicRpcSchema, ...AlchemyRpcSchema]>
>;

export type AlchemyTransport = AlchemyTransportBase & {
  updateHeaders(newHeaders: HeadersInit): void;
  config: AlchemyTransportConfig;
};

/**
 * Creates an Alchemy transport with the specified configuration options.
 * When sending all traffic to Alchemy, you must pass in one of rpcUrl, apiKey, or jwt.
 * If you want to send Bundler and Paymaster traffic to Alchemy and Node traffic to a different RPC, you must pass in alchemyConnection and nodeRpcUrl.
 *
 * @example
 * ### Basic Example
 * If the chain you're using is supported for both Bundler and Node RPCs, then you can do the following:
 * ```ts
 * import { alchemy } from "@account-kit/infra";
 *
 * const transport = alchemy({
 *  // NOTE: you can also pass in an rpcUrl or jwt here or rpcUrl and jwt
 *  apiKey: "your-api-key",
 * });
 * ```
 *
 * ### AA Only Chains
 * For AA-only chains, you need to specify the alchemyConnection and nodeRpcUrl since Alchemy only
 * handles the Bundler and Paymaster RPCs for these chains.
 * ```ts
 * import { alchemy } from "@account-kit/infra";
 *
 * const transport = alchemy({
 *  alchemyConnection: {
 *    apiKey: "your-api-key",
 *  },
 *  nodeRpcUrl: "https://zora.rpc.url",
 * });
 * ```
 *
 * @param {AlchemyTransportConfig} config The configuration object for the Alchemy transport.
 * @param {number} config.retryDelay Optional The delay between retries, in milliseconds.
 * @param {number} config.retryCount Optional The number of retry attempts.
 * @param {string} [config.alchemyConnection] Optional Alchemy connection configuration (if this is passed in, nodeRpcUrl is required).
 * @param {string} [config.fetchOptions] Optional fetch options for HTTP requests.
 * @param {string} [config.nodeRpcUrl] Optional RPC URL for node (if this is passed in, alchemyConnection is required).
 * @param {string} [config.rpcUrl] Optional RPC URL.
 * @param {string} [config.apiKey] Optional API key for Alchemy.
 * @param {string} [config.jwt] Optional JSON Web Token for authorization.
 * @returns {AlchemyTransport} The configured Alchemy transport object.
 */
export function alchemy(config: AlchemyTransportConfig): AlchemyTransport {
  const { retryDelay, retryCount } = config;
  // we create a copy here in case we create a split transport down below
  // we don't want to add alchemy headers to 3rd party nodes
  const fetchOptions = { ...config.fetchOptions };

  const connectionConfig = ConnectionConfigSchema.parse(
    config.alchemyConnection ?? config
  );

  const headersAsObject = convertHeadersToObject(fetchOptions.headers);

  // TODO: we probably should just log these headers during telemetry logging instead of doing this mutable header stuff
  fetchOptions.headers = {
    ...headersAsObject,
    "Alchemy-AA-Sdk-Version": VERSION,
  };

  if (connectionConfig.jwt != null) {
    fetchOptions.headers = {
      ...fetchOptions.headers,
      Authorization: `Bearer ${connectionConfig.jwt}`,
    };
  }

  const transport: AlchemyTransportBase = (opts) => {
    const { chain: chain_ } = opts;
    if (!chain_) {
      throw new ChainNotFoundError();
    }
    const chain = AlchemyChainSchema.parse(chain_);

    const rpcUrl =
      connectionConfig.rpcUrl == null
        ? `${chain.rpcUrls.alchemy.http[0]}/${connectionConfig.apiKey ?? ""}`
        : connectionConfig.rpcUrl;

    const innerTransport = (() => {
      if (config.alchemyConnection && config.nodeRpcUrl) {
        return split({
          overrides: [
            {
              methods: alchemyMethods,
              transport: http(rpcUrl, { fetchOptions }),
            },
          ],
          fallback: http(config.nodeRpcUrl, {
            fetchOptions: config.fetchOptions,
          }),
        });
      }

      return http(rpcUrl, { fetchOptions });
    })();

    return createTransport(
      {
        key: "alchemy",
        name: "Alchemy Transport",
        request: innerTransport(opts).request,
        retryCount: retryCount ?? opts?.retryCount,
        retryDelay,
        type: "alchemy",
      },
      { alchemyRpcUrl: rpcUrl, fetchOptions }
    );
  };

  return Object.assign(transport, {
    updateHeaders(newHeaders_: HeadersInit) {
      const newHeaders = convertHeadersToObject(newHeaders_);

      fetchOptions.headers = {
        ...fetchOptions.headers,
        ...newHeaders,
      };
    },
    config,
  });
}

const convertHeadersToObject = (
  headers?: HeadersInit
): Record<string, string> => {
  if (!headers) {
    return {};
  }

  if (headers instanceof Headers) {
    const headersObject = {} as Record<string, string>;
    headers.forEach((value, key) => {
      headersObject[key] = value;
    });
    return headersObject;
  }

  if (Array.isArray(headers)) {
    return headers.reduce((acc, header) => {
      acc[header[0]] = header[1];
      return acc;
    }, {} as Record<string, string>);
  }

  return headers;
};
