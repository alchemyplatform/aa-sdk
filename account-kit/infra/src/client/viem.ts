import { createBundlerClient } from "viem/account-abstraction";
import type { AlchemySmartAccountClientConfig } from "./smartAccountClient";

// Yea this isn't great or going to work well, we should just keep using our smart account client, but change the account type to match the viem one
export const foo = (params: AlchemySmartAccountClientConfig) => {
  const cl = createBundlerClient({
    transport: params.transport,
    chain: params.chain,
    paymasterContext: { policyId: params.policyId },
    paymaster: {
      async getPaymasterStubData(parameters) {
        if (!("policyId" in parameters.context)) {
          throw new Error("policyId is required");
        }

        throw new Error(
          "this is the biggest issue with viem's approach. we can't skip gas estimation during paymaster flows..."
        );
      },
      async getPaymasterData(parameters) {
        throw new Error("port over the paymaster data middleware");
      },
    },
    userOperation: {
      // This doesn't have the ability to skip gas estimation AFAICT
      estimateFeesPerGas(parameters) {
        throw new Error("port over the fee estimator middleware");
      },
    },
  });

  return cl;
};
