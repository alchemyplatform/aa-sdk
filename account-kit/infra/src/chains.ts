import type { Chain } from "viem";

import { defineChain } from "viem";
import {
  arbitrum as vab,
  arbitrumGoerli as vabg,
  arbitrumSepolia as vabs,
  base as vbase,
  baseGoerli as vbaseg,
  baseSepolia as vbases,
  fraxtal as vfrax,
  goerli as vgo,
  mainnet as vmain,
  optimism as vop,
  optimismGoerli as vopg,
  optimismSepolia as vops,
  polygon as vpg,
  polygonAmoy as vpga,
  polygonMumbai as vpgm,
  sepolia as vsep,
  arbitrumNova as vabn,
  zora as vzora,
  zoraSepolia as vzoras,
} from "viem/chains";

export type AlchemyChainConfig = {
  chain: Chain;
  rpcBaseUrl: string;
};

/**
 * Defines an Alchemy chain configuration by adding an Alchemy-specific RPC base URL to the chain's RPC URLs.
 *
 * @example
 * ```ts
 * import { defineAlchemyChain } from "@account-kit/infra";
 * import { sepolia } from "viem/chains";
 *
 * const chain = defineAlchemyChain({
 *  chain: sepolia,
 *  rpcBaseUrl: "https://eth-sepolia.g.alchemy.com/v2"
 * });
 * ```
 *
 * @param {AlchemyChainConfig} params The parameters for defining the Alchemy chain
 * @param {Chain} params.chain The original chain configuration
 * @param {string} params.rpcBaseUrl The Alchemy-specific RPC base URL
 * @returns {Chain} The updated chain configuration with the Alchemy RPC URL added
 */
export const defineAlchemyChain = ({
  chain,
  rpcBaseUrl,
}: {
  chain: Chain;
  rpcBaseUrl: string;
}): Chain => {
  return {
    ...chain,
    rpcUrls: {
      ...chain.rpcUrls,
      alchemy: {
        http: [rpcBaseUrl],
      },
    },
  };
};

export const arbitrum: Chain = {
  ...vab,
  rpcUrls: {
    ...vab.rpcUrls,
    alchemy: {
      http: ["https://arb-mainnet.g.alchemy.com/v2"],
    },
  },
};

export const arbitrumGoerli: Chain = {
  ...vabg,
  rpcUrls: {
    ...vabg.rpcUrls,
    alchemy: {
      http: ["https://arb-goerli.g.alchemy.com/v2"],
    },
  },
};

export const arbitrumSepolia: Chain = {
  ...vabs,
  rpcUrls: {
    ...vabs.rpcUrls,
    alchemy: {
      http: ["https://arb-sepolia.g.alchemy.com/v2"],
    },
  },
};
export const goerli: Chain = {
  ...vgo,
  rpcUrls: {
    ...vgo.rpcUrls,
    alchemy: {
      http: ["https://eth-goerli.g.alchemy.com/v2"],
    },
  },
};
export const mainnet: Chain = {
  ...vmain,
  rpcUrls: {
    ...vmain.rpcUrls,
    alchemy: {
      http: ["https://eth-mainnet.g.alchemy.com/v2"],
    },
  },
};
export const optimism: Chain = {
  ...vop,
  rpcUrls: {
    ...vop.rpcUrls,
    alchemy: {
      http: ["https://opt-mainnet.g.alchemy.com/v2"],
    },
  },
};
export const optimismGoerli: Chain = {
  ...vopg,
  rpcUrls: {
    ...vopg.rpcUrls,
    alchemy: {
      http: ["https://opt-goerli.g.alchemy.com/v2"],
    },
  },
};
export const optimismSepolia: Chain = {
  ...vops,
  rpcUrls: {
    ...vops.rpcUrls,
    alchemy: {
      http: ["https://opt-sepolia.g.alchemy.com/v2"],
    },
  },
};
export const sepolia: Chain = {
  ...vsep,
  rpcUrls: {
    ...vsep.rpcUrls,
    alchemy: {
      http: ["https://eth-sepolia.g.alchemy.com/v2"],
    },
  },
};
export const base: Chain = {
  ...vbase,
  rpcUrls: {
    ...vbase.rpcUrls,
    alchemy: {
      http: ["https://base-mainnet.g.alchemy.com/v2"],
    },
  },
};
export const baseGoerli: Chain = {
  ...vbaseg,
  rpcUrls: {
    ...vbaseg.rpcUrls,
    alchemy: {
      http: ["https://base-goerli.g.alchemy.com/v2"],
    },
  },
};
export const baseSepolia: Chain = {
  ...vbases,
  rpcUrls: {
    ...vbases.rpcUrls,
    alchemy: {
      http: ["https://base-sepolia.g.alchemy.com/v2"],
    },
  },
};

export const polygonMumbai: Chain = {
  ...vpgm,
  rpcUrls: {
    ...vpgm.rpcUrls,
    alchemy: {
      http: ["https://polygon-mumbai.g.alchemy.com/v2"],
    },
  },
};

export const polygonAmoy: Chain = {
  ...vpga,
  rpcUrls: {
    ...vpga.rpcUrls,
    alchemy: {
      http: ["https://polygon-amoy.g.alchemy.com/v2"],
    },
  },
};

export const polygon: Chain = {
  ...vpg,
  rpcUrls: {
    ...vpg.rpcUrls,
    alchemy: {
      http: ["https://polygon-mainnet.g.alchemy.com/v2"],
    },
  },
};

export const fraxtal: Chain = {
  ...vfrax,
  rpcUrls: {
    ...vfrax.rpcUrls,
  },
};

export const fraxtalSepolia: Chain = defineChain({
  id: 2523,
  name: "Fraxtal Sepolia",
  nativeCurrency: { name: "Frax Ether", symbol: "frxETH", decimals: 18 },
  rpcUrls: {
    default: {
      http: ["https://rpc.testnet-sepolia.frax.com"],
    },
  },
});

export const zora: Chain = {
  ...vzora,
  rpcUrls: {
    ...vzora.rpcUrls,
  },
};

export const zoraSepolia: Chain = {
  ...vzoras,
  rpcUrls: {
    ...vzoras.rpcUrls,
  },
};

export const worldChainSepolia: Chain = defineChain({
  id: 4801,
  name: "World Chain Sepolia",
  network: "World Chain Sepolia",
  nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
  rpcUrls: {
    default: {
      http: ["https://worldchain-sepolia.g.alchemy.com/v2"],
    },
    public: {
      http: ["https://worldchain-sepolia.g.alchemy.com/v2"],
    },
    alchemy: {
      http: ["https://worldchain-sepolia.g.alchemy.com/v2"],
    },
  },
});

export const worldChain: Chain = defineChain({
  id: 480,
  name: "World Chain",
  network: "World Chain",
  nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
  rpcUrls: {
    default: {
      http: ["https://worldchain-mainnet.g.alchemy.com/v2"],
    },
    public: {
      http: ["https://worldchain-mainnet.g.alchemy.com/v2"],
    },
    alchemy: {
      http: ["https://worldchain-mainnet.g.alchemy.com/v2"],
    },
  },
});

export const shapeSepolia: Chain = defineChain({
  id: 11011,
  name: "Shape Sepolia",
  network: "Shape Sepolia",
  nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
  rpcUrls: {
    default: {
      http: ["https://shape-sepolia.g.alchemy.com/v2"],
    },
    public: {
      http: ["https://shape-sepolia.g.alchemy.com/v2"],
    },
    alchemy: {
      http: ["https://shape-sepolia.g.alchemy.com/v2"],
    },
  },
});

export const shape: Chain = defineChain({
  id: 360,
  name: "Shape",
  network: "Shape",
  nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
  rpcUrls: {
    default: {
      http: ["https://shape-mainnet.g.alchemy.com/v2"],
    },
    public: {
      http: ["https://shape-mainnet.g.alchemy.com/v2"],
    },
    alchemy: {
      http: ["https://shape-mainnet.g.alchemy.com/v2"],
    },
  },
});

export const beraChainBartio: Chain = defineChain({
  id: 80084,
  name: "BeraChain Bartio",
  network: "BeraChain Bartio",
  nativeCurrency: { name: "Bera", symbol: "BERA", decimals: 18 },
  rpcUrls: {
    default: {
      http: ["https://berachain-bartio.g.alchemy.com/v2"],
    },
    public: {
      http: ["https://berachain-bartio.g.alchemy.com/v2"],
    },
    alchemy: {
      http: ["https://berachain-bartio.g.alchemy.com/v2"],
    },
  },
});

export const arbitrumNova: Chain = {
  ...vabn,
  rpcUrls: {
    ...vabn.rpcUrls,
  },
};
