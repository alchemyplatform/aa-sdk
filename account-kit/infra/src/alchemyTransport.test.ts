import * as AACoreModule from "@aa-sdk/core";
import { avalanche } from "viem/chains";
import { alchemy } from "./alchemyTransport.js";
import { sepolia } from "./chains.js";

describe("Alchemy Transport Tests", () => {
  it.each([
    { rpcUrl: "/api" },
    { jwt: "test" },
    { apiKey: "key" },
    { rpcUrl: "/api", jwt: "jwt" },
  ])("should successfully create a non-split transport", (args) => {
    expect(() =>
      alchemy({
        ...args,
      })
    ).not.toThrowError();
  });

  it.each([
    { rpcUrl: "/api" },
    { jwt: "test" },
    { apiKey: "key" },
    { rpcUrl: "/api", jwt: "jwt" },
  ])("should correctly create a split transport", (args) => {
    const splitSpy = vi.spyOn(AACoreModule, "split");
    alchemy({
      alchemyConnection: args,
      nodeRpcUrl: "/test",
    })({ chain: sepolia });

    expect(splitSpy.mock.calls.length).toBe(1);
    expect(splitSpy.mock.calls[0]).toMatchSnapshot();
  });

  it("should correctly do runtime validation when chain is not supported by Alchemy", () => {
    expect(() => alchemy({ rpcUrl: "/test" })({ chain: avalanche }))
      .toThrowErrorMatchingInlineSnapshot(`
      [ZodError: [
        {
          "code": "custom",
          "message": "chain must include an alchemy rpc url. See \`createAlchemyChain\` or import a chain from \`@account-kit/infra\`.",
          "fatal": true,
          "path": []
        }
      ]]
    `);
  });
});
