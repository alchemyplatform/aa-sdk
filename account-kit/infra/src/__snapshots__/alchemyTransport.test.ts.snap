// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Alchemy Transport Tests > should correctly create a split transport 1`] = `
[
  {
    "fallback": [Function],
    "overrides": [
      {
        "methods": [
          "eth_sendUserOperation",
          "eth_estimateUserOperationGas",
          "eth_getUserOperationReceipt",
          "eth_getUserOperationByHash",
          "eth_supportedEntryPoints",
          "rundler_maxPriorityFeePerGas",
          "pm_getPaymasterData",
          "pm_getPaymasterStubData",
        ],
        "transport": [Function],
      },
    ],
  },
]
`;

exports[`Alchemy Transport Tests > should correctly create a split transport 2`] = `
[
  {
    "fallback": [Function],
    "overrides": [
      {
        "methods": [
          "eth_sendUserOperation",
          "eth_estimateUserOperationGas",
          "eth_getUserOperationReceipt",
          "eth_getUserOperationByHash",
          "eth_supportedEntryPoints",
          "rundler_maxPriorityFeePerGas",
          "pm_getPaymasterData",
          "pm_getPaymasterStubData",
        ],
        "transport": [Function],
      },
    ],
  },
]
`;

exports[`Alchemy Transport Tests > should correctly create a split transport 3`] = `
[
  {
    "fallback": [Function],
    "overrides": [
      {
        "methods": [
          "eth_sendUserOperation",
          "eth_estimateUserOperationGas",
          "eth_getUserOperationReceipt",
          "eth_getUserOperationByHash",
          "eth_supportedEntryPoints",
          "rundler_maxPriorityFeePerGas",
          "pm_getPaymasterData",
          "pm_getPaymasterStubData",
        ],
        "transport": [Function],
      },
    ],
  },
]
`;

exports[`Alchemy Transport Tests > should correctly create a split transport 4`] = `
[
  {
    "fallback": [Function],
    "overrides": [
      {
        "methods": [
          "eth_sendUserOperation",
          "eth_estimateUserOperationGas",
          "eth_getUserOperationReceipt",
          "eth_getUserOperationByHash",
          "eth_supportedEntryPoints",
          "rundler_maxPriorityFeePerGas",
          "pm_getPaymasterData",
          "pm_getPaymasterStubData",
        ],
        "transport": [Function],
      },
    ],
  },
]
`;
